### DISCUSSION TITLE
forEach on optional state

### DISCUSSION THREAD

**Discussion by Rokas91** [pointfreeco/swift-composable-architecture#1624] [Q&A]
How to migrate this example using ReducerProtocol API?

```
enum AppAction: Equatable {
    
    case todo(id: UUID, action: TodoAction)
}

struct AppState {
    
    var todos: IdentifiedArrayOf<TodoState>?
}

todoReducer.
    .forEach(
        state: \.self,
        action: .self,
        environment: { $0 }
    )
    .optional()
    .pullback(
        state: \AppState.todos,
        action: /AppAction.todo,
        environment:  { _ in .init() }
    )
```

**Comment by tgrapperon** [comment]
Hey @Rokas91!, I'm a little late to the party, but this old unanswered question just popped on top of the conversations, so it may be worth providing an answer in any case someone stumbles upon it.
Without going into too much details, you need to drill down into the optional, and then in the forEach. Because of the way `ifLet` and `forEach` reducers are working using a trailing "modifier" style, it can be a little cumbersome in this setup, as you may introduce `EmptyReducer()` to serve as a basis for the transform. 
```swift
var body: some ReducerProtocolOf<Self> {
  EmptyReducer()
    ifLet(\.todos, action: .self) { 
      // A ReducerProtocolOf<IdentifiedArrayOf<TodoState, AppAction>>
      EmptyReducer()
        .forEach(\.self, action: /AppAction.todos) { // the action is some `case todos(TodoState.ID, TodoAction)`
          // A ReducerProtocolOf<TodoState, TodoAction>
          TodoReducer()
        }
    }
}
```
Something like this should probably work, but splitting the `Todos` domain into its own reducer could maybe make things clearer.
